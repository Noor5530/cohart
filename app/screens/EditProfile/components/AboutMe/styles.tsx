import { appFonts } from "components/text";
import { StyleSheet } from "react-native";
import { widthPercentageToDP } from "react-native-responsive-screen";

const styles = StyleSheet.create({
  fullName: {
    fontFamily: appFonts.InterSemiBold,
    fontWeight: "600",
    fontSize: 32,
    alignSelf: "center",
    marginTop: 25,
  },
  textArea: {
    borderWidth: 0,
    fontSize: 24,
    minHeight: 40,
    fontFamily: appFonts.InterRegular,
    marginBottom: 5,
    marginTop: 10
  },
  maxLength: {
    fontSize: 14,
    color: 'rgba(0,0,0,0.5)'
  },
  container: {
      paddingLeft: 16,
      paddingRight: 16
  },
  button: {
    width: widthPercentageToDP(30),
    height: 43,
    borderWidth: 1,
    borderColor: "black",
    justifyContent: "center",
    alignItems: "center",
    borderRadius: 4,
    alignSelf: "center",
    marginTop: 20,
    marginBottom: 20,
  },
  buttonText: {
    fontFamily: appFonts.InterBold,
    fontSize: 18,
  },
});

export default styles;
