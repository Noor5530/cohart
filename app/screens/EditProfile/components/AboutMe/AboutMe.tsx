import React, { useEffect, useRef, useState } from "react";
import { View, Text, TextInput, TouchableOpacity } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import ImagePicker from "react-native-image-crop-picker";
import { ActivityIndicator } from "react-native-paper";

import AppState from "models/reducers";

import AboutMeImages from "./components/AboutMeImages";
import { permissionAlert } from "lib/isIphone";
import styles from "./styles";
import { apiClient } from "services/client";
import saveProfileImage from "services/saveProfileImage";
import userProfileAboutMe from "services/userProfileAboutMe";
import RNFetchBlob from "rn-fetch-blob";
import { logOutRequest } from "store/actions/userActions";
import { LogOutRequestEnum } from "models/actions/user";

interface IProps {
  closeCollapsibleSection(): void;
  setSubmitForm(): void;
}

const AboutMe: React.FC<IProps> = ({
  closeCollapsibleSection,
  setSubmitForm,
}: IProps) => {
  const dispatch = useDispatch();
  const user = useSelector((state: AppState) => state.user);

  const [description, setDescription] = useState(user.bio || "");
  const [submitting, setSubmitting] = useState(false);
  const [imagePathOne, setImagePathOne] = useState(
    user.user_profile_images?.image1_original
  );
  const [imagePathTwo, setImagePathTwo] = useState(
    user.user_profile_images?.image2_original
  );
  const [imagePathThree, setImagePathThree] = useState(
    user.user_profile_images?.image3_original
  );
  const [imagePathFour, setImagePathFour] = useState(
    user.user_profile_images?.image4_original
  );
  const [imagePathFive, setImagePathFive] = useState(
    user.user_profile_images?.image5_original
  );
  const [imagePathSix, setImagePathSix] = useState(
    user.user_profile_images?.image6_original
  );
  const [imageLoading, setImageLoading] = useState<string[]>([]);
  const task = useRef();

  useEffect(() => {
    setDescription(user.bio || "");
    return () => {
      setDescription("");
    };
  }, [user]);

  const uploadImage = async (payload) => {
    try {
      setImageLoading((prev) => {
        if (!prev.includes(payload.location)) {
          return prev.concat(payload.location);
        }
        return prev;
      });
      apiClient
        .post("getSignedUrl", {
          file_name: payload.image_name,
          content_type: payload.image_type,
        })
        .then((data) => {
          task.current = RNFetchBlob.fetch(
            "PUT",
            data.data.data,
            {
              "Content-Type": payload.image_type,
            },
            RNFetchBlob.wrap(payload.uri)
          );
          return task.current.uploadProgress((written, total) => {
            console.log("Written", written);
            console.log("Total", total);
          });
        })
        .then((data) => {
          console.log(data, "========");
          saveImage(data.respInfo.redirects[0] as string, payload.location);
        })
        .catch((error) => {
          const { message } = error;
          console.log(message);
        });
    } catch (error: any) {
      if (error?.response?.status == 401) {
        dispatch(logOutRequest(LogOutRequestEnum.tokenExpire));
      }
      const { message } = error;
      console.log(message);
    }
  };

  const saveImage = async (imgUrl: string, location: string) => {
    const res = await saveProfileImage({
      id: user._id,
      image: imgUrl,
      image_location: location,
    });
    setImageLoading((prev) => {
      return prev.filter((item) => item !== location);
    });
    console.log("Response saved image => ", res);
  };

  const pickImages = (location) => () => {
    ImagePicker.openPicker({
      // width: 300,
      // height: 400,
      // cropping: true,
      mediaType: "photo",
      avoidEmptySpaceAroundImage: false,
    })
      .then((response) => {
        const image_name = response.path.substring(
          response.path.lastIndexOf("/") + 1
        );
        const image_type = response.mime;
        const uri = response.path;
        switch (location) {
          case 1:
            setImagePathOne(uri);
            break;
          case 2:
            setImagePathTwo(uri);
            break;
          case 3:
            setImagePathThree(uri);
            break;
          case 4:
            setImagePathFour(uri);
            break;
          case 5:
            setImagePathFive(uri);
            break;
          case 6:
            setImagePathSix(uri);
            break;
          default:
            break;
        }

        uploadImage({
          image_name,
          image_type,
          uri,
          location: String(location),
        });

        console.log("Response Image ==> " + response.path);
      })
      .catch((error: any) => {
        if (error?.message === "User did not grant library permission.") {
          permissionAlert();
        } else if (error?.message !== "User cancelled image selection") {
          // dispatch(enableSnackBar(error?.message));
        }
      });
  };

  const handleChangeText = (text: string) => {
    if (text.trim().length == 0 && text == " ") {
      return;
    }
    setDescription(text);
  };

  const handleUpdateBio = async () => {
    setSubmitting(true);
    try {
      await userProfileAboutMe({
        id: user._id,
        bio: description.replace(/\s+/g, " ").trim(),
      });
    } catch (error: any) {
      if (error?.response?.status == 401) {
        dispatch(logOutRequest(LogOutRequestEnum.tokenExpire));
      }
    }

    setSubmitting(false);
    setSubmitForm((prev) =>
      prev.map((item, index) => (index === 1 ? true : false))
    );
    closeCollapsibleSection();
  };

  return (
    <View>
      <AboutMeImages
        image2={imagePathTwo}
        image1={imagePathOne}
        image3={imagePathThree}
        image4={imagePathFour}
        image5={imagePathFive}
        image6={imagePathSix}
        pickImages={pickImages}
        imageLoading={imageLoading}
      />
      <View style={styles.container}>
        <Text style={styles.fullName}>{user?.full_name}</Text>
        <TextInput
          placeholder="Enter your bio.."
          multiline
          numberOfLines={18}
          style={styles.textArea}
          underlineColorAndroid={"transparent"}
          maxLength={345}
          onChangeText={handleChangeText}
          value={description}
        />

        <Text style={styles.maxLength}>
          {description.length}/345 characters
        </Text>

        <TouchableOpacity
          disabled={submitting}
          onPress={handleUpdateBio}
          style={styles.button}
        >
          {submitting ? (
            <ActivityIndicator color="black" size="small" />
          ) : (
            <Text style={styles.buttonText}>{"I'M DONE"}</Text>
          )}
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default AboutMe;
