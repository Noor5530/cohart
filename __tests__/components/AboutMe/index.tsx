import "react-native";
import React from "react";
import AboutMe from "../../../app/screens/EditProfile/components/AboutMe/AboutMe";
import renderer from "react-test-renderer";

jest.useFakeTimers();
jest.mock('rn-fetch-blob', () => ({}))

jest.mock("react-redux", () => ({
  useDispatch: () => {},
  useSelector: () => ({
    user: {
      affiliations: [],
      bio: "test",
      referral_by: {
        cover_image: "hnnj",
      },
    },
  }),
}));

test("renders correctly", () => {
  const tree = renderer
    .create(<AboutMe closeCollapsibleSection={jest.fn} setSubmitForm={jest.fn} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
