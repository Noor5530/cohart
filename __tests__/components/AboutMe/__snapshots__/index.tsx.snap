// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 200,
        "justifyContent": "center",
        "paddingTop": 20,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 200,
          "opacity": 1,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "width": undefined,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 200,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-5.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 20,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 61,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-2.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "left": undefined,
          "opacity": 1,
          "position": "absolute",
          "top": 40,
          "zIndex": -6,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 75,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-1.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 40,
          "left": undefined,
          "opacity": 1,
          "position": "absolute",
          "zIndex": -5,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 75,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-4.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 25,
          "opacity": 1,
          "position": "absolute",
          "right": -5,
          "zIndex": -5,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 75,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-3.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
          "position": "absolute",
          "right": undefined,
          "top": 50,
          "zIndex": -5,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "height": 75,
              "width": undefined,
            },
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "testUri": "../../../app/assets/Placeholder-6.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.6)",
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../app/assets/ic_edit.png",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingLeft": 16,
        "paddingRight": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "fontFamily": "Inter-SemiBold",
          "fontSize": 32,
          "fontWeight": "600",
          "marginTop": 25,
        }
      }
    />
    <TextInput
      allowFontScaling={true}
      maxLength={345}
      multiline={true}
      numberOfLines={18}
      onChangeText={[Function]}
      placeholder="Enter your bio.."
      rejectResponderTermination={true}
      style={
        Object {
          "borderWidth": 0,
          "fontFamily": "Inter-Regular",
          "fontSize": 24,
          "marginBottom": 5,
          "marginTop": 10,
          "minHeight": 40,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <Text
      style={
        Object {
          "color": "rgba(0,0,0,0.5)",
          "fontSize": 14,
        }
      }
    >
      0
      /345 characters
    </Text>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "borderColor": "black",
          "borderRadius": 4,
          "borderWidth": 1,
          "height": 43,
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
          "opacity": 1,
          "width": undefined,
        }
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "Inter-Bold",
            "fontSize": 18,
          }
        }
      >
        I'M DONE
      </Text>
    </View>
  </View>
</View>
`;
